client/allocrunner/taskrunner/artifact_hook.go

```diff
@@ -32,24 +32,39 @@ func (*artifactHook) Name() string {
 }
 
 func (h *artifactHook) Prestart(ctx context.Context, req *interfaces.TaskPrestartRequest, resp *interfaces.TaskPrestartResponse) error {
+	var (
+		isDone = true
+	)
+
 	if len(req.Task.Artifacts) == 0 {
-		resp.Done = true
+		resp.Done = isDone
 		return nil
 	}
 
-	// Initialize hook state to store download progress
+	// Initialize hook state to store download progress.
 	resp.State = make(map[string]string, len(req.Task.Artifacts))
 
-	h.eventEmitter.EmitEvent(structs.NewTaskEvent(structs.TaskDownloadingArtifacts))
-
 	for _, artifact := range req.Task.Artifacts {
 		aid := artifact.Hash()
-		if req.PreviousState[aid] != "" {
+		if req.PreviousState[aid] != "" && !artifact.AlwaysFetch {
 			h.logger.Trace("skipping already downloaded artifact", "artifact", artifact.GetterSource)
 			resp.State[aid] = req.PreviousState[aid]
 			continue
 		}
 
+		// If there are any artifacts that should always be fetched before starting a task,
+		// then this hook must rerun on restarts, so it can't be marked as done.
+		if artifact.AlwaysFetch {
+			isDone = false
+		}
+
+		// Emit the event based on the flag.
+		if artifact.AlwaysFetch && req.PreviousState[aid] != "" {
+			h.eventEmitter.EmitEvent(structs.NewTaskEvent(structs.TaskReDownloadingArtifacts))
+		} else {
+			h.eventEmitter.EmitEvent(structs.NewTaskEvent(structs.TaskDownloadingArtifacts))
+		}
```

Comment:
Unfortunately I think this changes how often we emit task events for artifacts from `1 task event ever` to `1 per artifact + 1 per redownload` on restarts.

I think the behavior would ideally be:

1. Maintain existing behavior of `1 task event ever` if no artifacts have `always_fetch=true`
2. Emit 1 additional event *per restart* if *any* artifact has `always_fetch=true`

Unfortunately I think this means refactoring this block of code. My main concern is that we only store 10 task events, so if a task has a lot of artifacts they may miss other events.